lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -740,6 +797,88 @@ ParseResult FIRParser::parseOptionalRUW(RUWAttr &result) {
   return success();
 }
 
+ParseResult FIRParser::importAnnotations(const SMLoc &loc,
+                                         StringRef annotationsStr,
+                                         MLIRContext *context,
+                                         StringRef errorMsgParse) {
+
+  auto annotations = llvm::json::parse(annotationsStr);
+  if (auto err = annotations.takeError()) {
+    handleAllErrors(std::move(err), [&](const llvm::json::ParseError &a) {
+      auto diag = emitError(loc, errorMsgParse);
+      diag.attachNote() << a.message();
+    });
+    return failure();
+  }
+
+  llvm::json::Path::Root root;
+  llvm::StringMap<ArrayAttr> annotationMap;
+  if (annotations) {
+    if (!fromJSON(annotations.get(), annotationMap, root, context)) {
+      auto diag = emitError(loc, "Invalid/unsupported annotation format");
+      std::string jsonErrorMessage =
+          "See inline comments for problem area in JSON:\n";
+      llvm::raw_string_ostream s(jsonErrorMessage);
+      root.printErrorContext(annotations.get(), s);
+      diag.attachNote() << jsonErrorMessage;
+      return failure();
+    }
+  }
+
+  for (auto a : annotationMap.keys()) {
+    if (!state.annotationMap[a]) {
+      state.annotationMap[a] = annotationMap[a];
+      continue;
+    }
+
+    SmallVector<Attribute> annotationVec;
+    for (auto a : state.annotationMap[a])
+      annotationVec.push_back(a);
+    for (auto a : annotationMap[a])
+      annotationVec.push_back(a);
+    state.annotationMap[a] = ArrayAttr::get(context, annotationVec);
+  }
+
+  return success();
+}
+
+ParseResult FIRParser::importAnnotationFile(const SMLoc &loc,
+                                            MLIRContext *context) {
+
+  ParseResult result = success();
+  if (state.annotationsBuf)
```

Comment:
I'd recomment inverting + early exit to simplify the logic:

`if (!state.annotationsBuf) return success();`

Reply:
Fixed in a later commit.

Reply:
Fixed in a later commit.